### PR TITLE
[Merged by Bors] - feat(Data/List/Chain): well-foundedness of chains under `List.Lex`

### PR DESCRIPTION
**Pull request by alreadydone**
shorter proof of the main result of #6361

Zulip: https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/List.2ELex.20is.20well.20founded.20on.20decreasing.20lists/near/382708025

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by dagurtomas** [comment]
Maybe it would be instructive to add some of what you wrote in [this message](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/List.2ELex.20is.20well.20founded.20on.20decreasing.20lists/near/382766415) as comments inside the proof?

File: Mathlib/Data/List/Chain.lean [hunk 1]
```diff
@@ -440,3 +440,50 @@ theorem relationReflTransGen_of_exists_chain (l : List α) (hl₁ : Chain r a l)
 #align list.relation_refl_trans_gen_of_exists_chain List.relationReflTransGen_of_exists_chain
 
 end List
+
+
+/-! In this section, we consider the type of `r`-decreasing chains (`List.Chain' (flip r)`)
+  equipped with lexicographic order `List.Lex r`. -/
+
+variable {α : Type*} {r : α → α → Prop}
```
Comment (ocfnash) [inline Mathlib/Data/List/Chain.lean:448]: What do you think about introducing an abbreviation for the repeating terms appearing below:
```suggestion
variable {α : Type*} (r : α → α → Prop)

abbrev List.chains := { l : List α // l.Chain' (flip r) }

abbrev List.lex_chains : List.chains r → List.chains r → Prop := fun l m ↦ List.Lex r l.val m.val

variable {r}
```
(and using them in the three statements below)

File: Mathlib/Data/List/Chain.lean [hunk 2]
```diff
@@ -440,3 +440,50 @@ theorem relationReflTransGen_of_exists_chain (l : List α) (hl₁ : Chain r a l)
...
+
+/-- If an `r`-decreasing chain `l` is empty or its head is accessible by `r`, then
+  `l` is accessible by the lexicographic order `List.Lex r`. -/
+theorem Acc.list_chain' {l : {l : List α // l.Chain' (flip r)}} (acc : ∀ a ∈ l.val.head?, Acc r a) :
+    Acc (fun l m ↦ List.Lex r l.val m.val) l := by
```
Comment (ocfnash) [inline Mathlib/Data/List/Chain.lean:453]:
```suggestion
theorem Acc.list_chain' {l : List.chains r} (acc : ∀ a ∈ l.val.head?, Acc r a) :
    Acc (List.lex_chains r) l := by
```

File: Mathlib/Data/List/Chain.lean [hunk 3]
```diff
@@ -440,3 +440,50 @@ theorem relationReflTransGen_of_exists_chain (l : List α) (hl₁ : Chain r a l)
...
+      intro hl
+      apply Acc.intro
+      rintro ⟨_ | ⟨b, m⟩, hm⟩ (_ | hr | hr)
+      · apply Acc.intro; rintro ⟨_⟩ ⟨_⟩
+      · apply ihl ⟨m, (List.chain'_cons'.1 hm).2⟩ hr
+      · apply ih b hr
+
+/-- If `r` is well-founded, the lexicographic order on `r`-decreasing chains is also. -/
+theorem WellFounded.list_chain' (hwf : WellFounded r) :
+    @WellFounded {l : List α // l.Chain' (flip r)} (fun l m ↦ List.Lex r l.val m.val) :=
```
Comment (ocfnash) [inline Mathlib/Data/List/Chain.lean:484]:
```suggestion
    WellFounded (r := List.lex_chains r) :=
```

File: Mathlib/Data/List/Chain.lean [hunk 4]
```diff
@@ -440,3 +440,50 @@ theorem relationReflTransGen_of_exists_chain (l : List α) (hl₁ : Chain r a l)
...
+  ⟨fun _ ↦ Acc.list_chain' (fun _ _ => hwf.apply _)⟩
+
+instance [hwf : IsWellFounded α r] :
+    IsWellFounded {l : List α // l.Chain' (flip r)} (fun l m ↦ List.Lex r l.val m.val) :=
```
Comment (ocfnash) [inline Mathlib/Data/List/Chain.lean:488]:
```suggestion
    IsWellFounded (List.chains r) (List.lex_chains r) :=
```

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/Data/List/Chain.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/Data/List/Chain.lean:448]: Good idea!

**Comment by alreadydone** [comment]
Thanks!

bors r+

**Pull request closed without merging**